File: scripts/audit-book-pdf.mjs
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';
import { PdfReader } from 'pdfjs-dist/legacy/build/pdf.mjs';

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const ROOT = path.join(__dirname, '..');
const BOOK = path.join(ROOT, 'book');

async function extractPdfText(filePath) {
  const data = new Uint8Array(fs.readFileSync(filePath));
  const doc = await PdfReader({ data }).promise;
  let text = '';
  for (let i = 1; i <= doc.numPages; i++) {
    const page = await doc.getPage(i);
    const content = await page.getTextContent();
    text += content.items.map((it) => it.str).join(' ') + '\n';
  }
  return text;
}

function readUtf8(p) {
  return fs.readFileSync(p, 'utf8');
}

const issues = [];

function add(cat, item, detail) {
  issues.push({ cat, item, detail });
}

const pdfs = fs.readdirSync(BOOK).filter((f) => f.endsWith('.pdf'));
const pdfTexts = {};
for (const name of pdfs) {
  try {
    pdfTexts[name] = await extractPdfText(path.join(BOOK, name));
  } catch (e) {
    pdfTexts[name] = '';
    add('PDF 읽기', name, String(e.message || e));
  }
}

const dictionary = readUtf8(path.join(ROOT, 'js', 'technology', 'dictionary.js'));
const content = readUtf8(path.join(ROOT, 'js', 'technology', 'content-data.js'));
const index = readUtf8(path.join(ROOT, 'index.html'));
const termGuide = readUtf8(path.join(BOOK, 'KDS-KCS_용어기준.md'));
const imagesJs = readUtf8(path.join(ROOT, 'js', 'technology', 'images.js'));

// --- KDS tunnel items vs tree ---
const kdsTunnelRequired = [
  '지표 및 지중침하',
  '내공변위',
  '천단침하',
  '지중변위',
  '막장전방 선행변위',
  '록볼트 축력',
  '숏크리트 응력',
  '강지보 응력'
];
const tunnelSection = dictionary.split('fields/tunnel')[1]?.split('fields/bridge')[0] || '';
const tunnelChildLabels = [...tunnelSection.matchAll(/label: '([^']+)'/g)].map((m) => m[1]);

for (const term of kdsTunnelRequired) {
  if (term === '지표 및 지중침하') {
    if (!tunnelChildLabels.some((l) => l.includes('지표') || l.includes('지중침하'))) {
      add('KDS 4.1.5 터널 항목 누락', term, '트리에 전용 하위 메뉴 없음 (카테고리 overview에만 언급)');
    }
  }
  if (term === '막장전방 선행변위') {
    if (!tunnelChildLabels.some((l) => l.includes('선행') || l.includes('막장'))) {
      add('KDS 4.1.5 터널 항목 누락', term, '트리·리프 페이지 없음');
    }
  }
  if (term === '강지보 응력') {
    if (!content.includes('강지보')) {
      add('KDS 4.1.5 터널 항목 누락', term, 'KDS 4.1.5.2에 명시, 웹 콘텐츠 미반영');
    }
  }
}

// slug convergence
if (!dictionary.includes("id: 'fields/tunnel/convergence', label: '내공변위'")) {
  add('용어/slug', 'fields/tunnel/convergence', '화면 라벨이 내공변위가 아닐 수 있음');
}

// KDS 27 50 10 in book but not in term guide
if (pdfTexts['KDS 27 50 10 터널 계측(23.09).pdf'] && !termGuide.includes('KDS 27 50 10')) {
  add('기준서 참조 불일치', 'KDS 27 50 10', 'book PDF 존재, KDS-KCS_용어기준.md는 KDS 11 10 15만 상위 명시');
}

// Bridge KCS
const bridgePdf = pdfTexts['KCS 24 99 05 교량계측시설(23.09).pdf'] || '';
const bridgeSection = dictionary.split('fields/bridge')[1]?.split('fields/slope')[0] || '';
const bridgeLabels = [...bridgeSection.matchAll(/label: '([^']+)'/g)].map((m) => m[1]);
const kcsBridgeHints = ['교량 처짐', '교량 횡변위', '교량 종방향', '온도', '지진', '교량 기초'];
for (const h of kcsBridgeHints) {
  if (bridgePdf.includes(h.replace('교량 ', '')) || bridgePdf.includes(h)) {
    if (!bridgeLabels.some((l) => h.includes(l) || l.includes(h.replace('교량 ', '')))) {
      if (['온도', '지진'].includes(h) && bridgePdf.includes(h)) {
        add('KCS 24 99 05 교량 항목 미분리', h, 'KCS에 언급되나 트리 하위 항목 없음');
      }
    }
  }
}

// Dam KCS
const damPdf = pdfTexts['KCS 54 20 25 댐 계측설비(18.08).pdf'] || '';
const damSection = dictionary.split('fields/dam')[1]?.split('group-sensor')[0] || '';
const damLabels = [...damSection.matchAll(/label: '([^']+)'/g)].map((m) => m[1]);
for (const h of ['온도', '지진', '변형률', '유량', '기울기']) {
  if (damPdf.includes(h) && !damLabels.some((l) => l.includes(h))) {
    add('KCS 54 20 25 댐 항목 미분리', h, `현재 트리: ${damLabels.join(', ')}`);
  }
}

// Forbidden / risky terms in source mjs
const scanDirs = [path.join(ROOT, 'scripts', 'content-data')];
const sourceFiles = [];
function walk(d) {
  for (const n of fs.readdirSync(d)) {
    const f = path.join(d, n);
    if (fs.statSync(f).isDirectory()) walk(f);
    else if (n.endsWith('.mjs')) sourceFiles.push(f);
  }
}
walk(scanDirs[0]);
sourceFiles.push(path.join(ROOT, 'js', 'technology', 'content-data.js'));

const forbidden = [
  [/내공변위[^\\n]{0,20}\(수렴\)/, '내공변위(수렴)'],
  [/1·2·3차 경보/, '1·2·3차 경보'],
  [/1차 관리기준/, '1차 관리기준'],
  [/인클리노미터/, '인클리노미터 동일시'],
  [/진동현식/, '진동현식']
];
for (const f of sourceFiles) {
  const rel = path.relative(ROOT, f);
  const lines = readUtf8(f).split(/\r?\n/);
  lines.forEach((line, i) => {
    if (line.includes('terminology-ok')) return;
    for (const [re, name] of forbidden) {
      if (re.test(line)) add('금지 용어', name, `${rel}:${i + 1}`);
    }
  });
}

// settlement gauge table '수렴' as stage label
if (/\['잔류',\s*'수렴'/.test(content)) {
  add('KDS 용어 주의', '수렴 (표 헤더)', '침하계 데이터 표에 단독 「수렴」열 — KDS 의미(계측값 수렴)와 혼동 가능');
}

// LVDT in images
if (imagesJs.includes('LVDT')) {
  add('이미지 파일명', 'LVDT', 'IMG-040 파일명에 LVDT — 화면 용어는 변위계여야 함');
}

// JSON-LD address
if (index.includes('4층 403호') && index.includes('"streetAddress": "가산디지털1로 84, 403호"')) {
  add('회사정보 불일치', '본사 주소', '본문·연락처: 4층 403호 / JSON-LD: 층 없이 403호만');
}

// Company brochures
for (const name of ['25년 12월 지명원-(주)신계측기술정보.pdf', '241226 지명원_신계측기술정보.pdf']) {
  const t = pdfTexts[name] || '';
  if (!t) continue;
  const checks = [
    ['황인섭', '대표이사'],
    ['E-09-006550', '엔지니어링사업자'],
    ['108-81-66678', '사업자등록번호'],
    ['가산디지털', '본사 주소'],
    ['복용동로', '지사 주소'],
    ['865', '전화']
  ];
  for (const [needle, label] of checks) {
    if (t.includes(needle) && !index.includes(needle.replace('865', '865'))) {
      if (needle === '865' && index.includes('865-2188')) continue;
      add('지명원 vs 홈페이지', label, `${name}에 있으나 index.html 검색 불일치 가능`);
    }
  }
  // ISO / patents from brochure
  if (/ISO\s*9001|ISO9001/i.test(t) && !/ISO/i.test(index)) {
    add('지명원 vs 홈페이지', 'ISO 인증', `${name}에 ISO 언급, index 인증 섹션에 ISO 미표기`);
  }
  if (/특허\s*\d|제\d+호/.test(t)) {
    const patentNums = [...t.matchAll(/제\s*(\d{6,})호/g)].map((m) => m[1]);
    const missing = patentNums.filter((n) => !index.includes(n));
    if (missing.length) {
      add('지명원 vs 홈페이지', '특허 번호', `지명원 특허 ${missing.slice(0, 3).join(', ')}… index 미반영`);
    }
  }
}

// Boilerplate
const bp = '인접 센서(지중경사계, 하중계, 지하수위계, 간극수압계 등)와의 동시 변화';
const bpCount = (content.match(new RegExp(bp.replace(/[()]/g, '\\$&'), 'g')) || []).length;
if (bpCount > 10) {
  add('콘텐츠 품질', '보일러플레이트', `분야별 페이지에 동일 문단 ${bpCount}회 반복 — KDS/KCS 맞춤 서술 부족`);
}

// KCS ground vs website scope
const kcsGround = pdfTexts['KCS 11 10 15 시공 중 지반계측_(25. 12. 24).pdf'] || '';
if (kcsGround.includes('계측책임자') && !content.includes('계측책임자')) {
  add('KCS 용어 누락', '계측책임자', 'KCS 1.3에 정의, 기술자료 본문에 거의 미사용');
}

// inclinometer dual path
add('구조/운영', '지중경사계 이중 콘텐츠', '/homepage/sensors/inclinometer/ 정적 페이지 + SPA — 동기화·canonical 분리');

// validate term guide vs book PDF count
if (!fs.existsSync(path.join(BOOK, 'KDS 11 10 15 지반계측(25.12).pdf'))) {
  add('book 자산', 'KDS PDF', '파일 누락');
}

const outPath = path.join(ROOT, 'docs', 'book-consistency-audit.json');
fs.writeFileSync(outPath, JSON.stringify({ generated: new Date().toISOString(), issue_count: issues.length, issues }, null, 2), 'utf8');
console.log('Wrote', outPath);
console.log('Issues:', issues.length);
issues.forEach((i) => console.log(`- [${i.cat}] ${i.item}: ${i.detail}`));

File: scripts/audit_book_pdf.py
# -*- coding: utf-8 -*-
import json, re, os, sys
from pathlib import Path
from pypdf import PdfReader

if hasattr(sys.stdout, 'reconfigure'):
    sys.stdout.reconfigure(encoding='utf-8', errors='replace')

ROOT = Path(r'x:\website\homepage')
BOOK = ROOT / 'book'
issues = []

def add(cat, item, detail):
    issues.append({'cat': cat, 'item': item, 'detail': detail})

def leaf_chunk(content: str, node_id: str, limit: int = 15000) -> str:
    """LEAVES block in content-data.js (not sources-only string entry)."""
    marker = f'"{node_id}": {{'
    idx = content.find(marker)
    if idx < 0:
        return ''
    return content[idx : idx + limit]


def png_path_for(content: str, img_id: str) -> str:
    m = re.search(rf'"{re.escape(img_id)}"[^}}]*?"png":\s*"([^"]+)"', content, re.S)
    return m.group(1) if m else ''


def read_utf8(p):
    return Path(p).read_text(encoding='utf-8')

pdf_texts = {}
for p in sorted(BOOK.glob('*.pdf')):
    try:
        r = PdfReader(str(p))
        pdf_texts[p.name] = '\n'.join((page.extract_text() or '') for page in r.pages)
    except Exception as e:
        pdf_texts[p.name] = ''
        add('PDF 읽기', p.name, str(e))

dictionary = read_utf8(ROOT / 'js/technology/dictionary.js')
content = read_utf8(ROOT / 'js/technology/content-data.js')
index = read_utf8(ROOT / 'index.html')
term_guide = read_utf8(BOOK / 'KDS-KCS_용어기준.md')
images_js = read_utf8(ROOT / 'js/technology/images.js')

tunnel_section = dictionary.split('fields/tunnel')[1].split('fields/bridge')[0]
tunnel_child_labels = re.findall(r"label: '([^']+)'", tunnel_section)

kds_tunnel_required = [
    ('지표 및 지중침하', 'surface-subsidence'),
    ('막장전방 선행변위', 'face-advance'),
    ('강지보 응력', 'steel-support'),
]
for term, slug in kds_tunnel_required:
    if f'fields/tunnel/{slug}' not in dictionary:
        add('KDS 4.1.5 터널 항목 누락', term, f'트리 리프 fields/tunnel/{slug} 없음')

if 'fields/tunnel/blast-vibration' not in dictionary:
    add('KDS 4.1.5 선택', '발파진동', 'fields/tunnel/blast-vibration 리프 없음')

if 'fields/dam/river-levee' not in dictionary:
    add('KDS 제방', '하천제방', 'fields/dam/river-levee 리프 없음')

if 'fields/harbor' not in dictionary:
    add('KDS 4.1.8', '항만·호안', 'fields/harbor 트리 없음')

if 'fields/building' not in dictionary:
    add('KCS 3.9', '건축공사', 'fields/building 트리 없음')

if pdf_texts.get('KDS 27 50 10 터널 계측(23.09).pdf') and 'KDS 27 50 10' not in term_guide:
    add('기준서 참조 불일치', 'KDS 27 50 10', 'book PDF 존재, KDS-KCS_용어기준.md 미명시')

bridge_pdf = pdf_texts.get('KCS 24 99 05 교량계측시설(23.09).pdf', '')
dam_pdf = pdf_texts.get('KCS 54 20 25 댐 계측설비(18.08).pdf', '')

def tree_block(dictionary: str, field_id: str, next_field_id: str) -> str:
    """Extract TREE children block between two field category ids."""
    m = re.search(
        rf"id: '{re.escape(field_id)}'[\s\S]*?id: '{re.escape(next_field_id)}'",
        dictionary,
    )
    return m.group(0) if m else ''

bridge_section = tree_block(dictionary, 'fields/bridge', 'fields/slope')
for slug in ['temperature', 'seismic', 'expansion-joint']:
    if bridge_pdf and f'fields/bridge/{slug}' not in bridge_section:
        add('KCS 24 99 05 교량 항목 미분리', slug, 'KCS 대비 트리 리프 없음')

dam_section = tree_block(dictionary, 'fields/dam', 'fields/harbor')
for slug in ['temperature', 'seismic', 'strain', 'tilt']:
    if dam_pdf and f'fields/dam/{slug}' not in dam_section:
        add('KCS 54 20 25 댐 항목 미분리', slug, 'KCS 대비 트리 리프 없음')
# 유량은 fields/dam/leakage(누수)에서 다룸

source_files = list((ROOT / 'scripts/content-data').rglob('*.mjs'))
source_files += [
    ROOT / 'js/technology/content-data.js',
    ROOT / 'index.html',
    ROOT / 'sensors/inclinometer/index.html',
]
forbidden = [
    (r'내공변위[^\\n]{0,20}\(수렴\)', '내공변위(수렴)'),
    (r'1·2·3차 경보', '1·2·3차 경보'),
    (r'1차 관리기준', '1차 관리기준'),
    (r'인클리노미터', '인클리노미터 동일시'),
    (r'진동현식', '진동현식'),
]
for f in source_files:
    rel = str(f.relative_to(ROOT))
    lines = read_utf8(f).splitlines()
    for i, line in enumerate(lines, 1):
        if 'terminology-ok' in line:
            continue
        for pat, name in forbidden:
            if re.search(pat, line):
                add('금지 용어', name, f'{rel}:{i}')

if re.search(r"\['잔류',\s*'수렴'", content):
    add('KDS 용어 주의', '수렴 (표 헤더)', '침하계 데이터 표에 단독 「수렴」열 — KDS 의미와 혼동 가능')

if 'LVDT' in png_path_for(images_js, 'IMG-040').upper():
    add('이미지 파일명', 'LVDT', 'IMG-040 PNG 경로에 LVDT 잔존')

if '"streetAddress": "가산디지털1로 84, 403호"' in index and '4층 403호' in index:
    if '"streetAddress": "가산디지털1로 84, 4층 403호"' not in index:
        add('회사정보 불일치', '본사 주소', 'JSON-LD streetAddress에 4층 누락')

for name in ['25년 12월 지명원-(주)신계측기술정보.pdf', '241226 지명원_신계측기술정보.pdf']:
    t = pdf_texts.get(name, '')
    if not t:
        continue
    if re.search(r'ISO\s*9001|ISO9001', t, re.I) and not re.search(r'ISO', index):
        add('지명원 vs 홈페이지', 'ISO 인증', f'{name}에 ISO 언급, index 인증 섹션에 ISO 미표기')
    patent_nums = [m.group(1) for m in re.finditer(r'제\s*(\d{6,})\s*호', t)]
    missing = [n for n in patent_nums if n not in index]
    if missing:
        add('지명원 vs 홈페이지', '특허 번호', f'{name}: {", ".join(missing[:5])} index 미반영')

bp = '인접 센서(지중경사계, 하중계, 지하수위계, 간극수압계 등)와의 동시 변화'
bp_count = content.count(bp)
if bp_count > 10:
    add('콘텐츠 품질', '보일러플레이트', f'동일 문단 {bp_count}회 반복')

kcs_ground = pdf_texts.get('KCS 11 10 15 시공 중 지반계측_(25. 12. 24).pdf', '')
if '계측책임자' in kcs_ground and '계측책임자' not in content:
    add('KCS 용어 누락', '계측책임자', 'KCS 1.3 정의, 기술자료 본문에 거의 미사용')

gnss_pdf = pdf_texts.get('GNSS.pdf', '')
if gnss_pdf:
    gnss_chunk = leaf_chunk(content, 'sensors/gnss')
    for kw in ('기준국', '이동국', 'RTK', 'GNSS', 'GPS'):
        if kw in gnss_pdf and kw not in gnss_chunk:
            add('GNSS.pdf vs 웹', kw, 'book/GNSS.pdf 키워드 — sensors/gnss 본문 미포함')
    if '/homepage/book/GNSS.pdf' not in gnss_chunk:
        add('GNSS.pdf vs 웹', 'PDF 링크', 'sensors/gnss detailLink·본문 PDF href 누락')
    if 'IMG-043' not in gnss_chunk:
        add('GNSS.pdf vs 웹', 'IMG-043', 'sensors/gnss hero Figure 누락')

# 지중경사계: SPA 요약 + 정적 상세 (의도적 이중 경로). SEO canonical = 정적 페이지.
incl_static = ROOT / 'sensors/inclinometer/index.html'
incl_seo_dup = ROOT / 'technology/sensors/inclinometer/index.html'
sitemap = read_utf8(ROOT / 'sitemap.xml') if (ROOT / 'sitemap.xml').exists() else ''
incl_ok = (
    incl_static.exists()
    and 'detailLink' in content
    and '/homepage/sensors/inclinometer/' in content
    and not incl_seo_dup.exists()
    and '/homepage/sensors/inclinometer/' in sitemap
)
if not incl_ok:
    add('구조/운영', '지중경사계 이중 콘텐츠', 'SPA+정적 canonical 정책 미충족 — detailLink·SKIP_IDS·sitemap 점검')

# Extract brochure business areas
for name in ['25년 12월 지명원-(주)신계측기술정보.pdf']:
    t = pdf_texts.get(name, '')
    if not t:
        continue
    for kw in ['지반조사', '재하시험', '구조안전', '계측관리', '자동화', 'AI']:
        if kw in t and kw not in index:
            add('지명원 vs 홈페이지', f'사업 키워드 {kw}', f'지명원에 있으나 index.html 본문 검색 미매칭')

# KDS shoring - earth pressure vs load cell
if '로드셀' in read_utf8(BOOK / '_kds_kcs_term_extract.json'):
    chunk = leaf_chunk(content, 'fields/retaining-excavation/anchor', 8000)
    if chunk and '로드셀' not in chunk:
        add('KDS 용어', '어스앵커 장력', 'KDS 표4.1-1 앵커=로드셀, 어스앵커 페이지에 로드셀 병기 부족')

# Check certifications section content vs brochure
cert_section = index[index.find('등록·인증·특허'):index.find('등록·인증·특허')+3000] if '등록·인증·특허' in index else ''
if 'SW' in cert_section and 'GS' in t if (t := pdf_texts.get('25년 12월 지명원-(주)신계측기술정보.pdf','')) else False:
    pass

out = ROOT / 'docs' / 'book-consistency-audit.json'
out.write_text(json.dumps({'issue_count': len(issues), 'issues': issues}, ensure_ascii=False, indent=2), encoding='utf-8')
print('Issues:', len(issues))
for i in issues:
    print(f"[{i['cat']}] {i['item']}: {i['detail']}")
